src/AggregateFunctions/QuantileExact.h

```diff
@@ -64,18 +70,45 @@ struct QuantileExact
         buf.read(reinterpret_cast<char *>(array.data()), size * sizeof(array[0]));
     }
 
-    /// Get the value of the `level` quantile. The level must be between 0 and 1.
-    Value get(Float64 level)
+    size_t getElementNumber(Float64 level) const
+    {
+        return level < 1 ? level * array.size()
+                         : (array.size() - 1);
+    }
+
+    void finalize(Float64 level)
     {
         if (!array.empty())
         {
-            size_t n = level < 1
-                ? level * array.size()
-                : (array.size() - 1);
-
+            size_t n = getElementNumber(level);
             std::nth_element(array.begin(), array.begin() + n, array.end());    /// NOTE You can think of the radix-select algorithm.
-            return array[n];
         }
+    }
+
+    void finalize(const Float64 * levels, const size_t * indices, size_t size)
+    {
+        if (!array.empty())
+        {
+            size_t interval_start = 0;
+            for (size_t i = 0; i < size; ++i)
+            {
+                auto level = levels[indices[i]];
+                size_t n = getElementNumber(level);
+
+                if (n + 1 == interval_start)
+                    continue;
+
+                std::nth_element(array.begin() + interval_start, array.begin() + n, array.end());
```

Comment:
So in this function we just pre-select the array elements we need? But why? Does it give performance benefit on some query, or is it somehow required for correctness?

Reply:
> So in this function we just pre-select the array elements we need? 

Yes. It is done do avoid changing internal state (sorting an array) at the time we get final result (in `IAggregateFunction::insertResultInto`).
It doesn't give any performance (except we run `insertResultInto` several times).
But if we change internal state in `insertResultInto`, it may cause race.

Reply:
Understood. The list of broken functions that const_cast and change the data in `insertResultInto`: https://github.com/ClickHouse/ClickHouse/pull/10383#issuecomment-627340591